Clear activeTab and action tab customization when tab navigates.

xeenon opened the pull request:
#### dfe319166e75f999730302ef7d9309a21a6ac1cd
<pre>
Clear activeTab and action tab customization when tab navigates.
<a href="https://webkit.org/b/263918">https://webkit.org/b/263918</a>
<a href="https://rdar.apple.com/problem/117704349">rdar://problem/117704349</a>

Reviewed by Brian Weinstein.

Moves the webNavigation load events to WebExtensionContext, since more things will
need to tie into these hooks.

Fixes injected style removal when webPageID is used, so it only removes the styles
when the page they apply to is navigated. Also this was incorrectly behind the
webNavigation permission checks before, when it needs to always happen.

Properly clears the activeTb permission on tab navigation, and fires the onUpdated
events for URL and Title when activeTab comes and goes, as well as any pattern grant.
This matches what Safari currently does when permissions are granted.

Clears the action customizations if the tab navigates. This matches the other browsers
and what Safari currently does.

Fixes the match pattern notifications to use the right constant.

* Source/WebKit/UIProcess/Extensions/Cocoa/WebExtensionActionCocoa.mm:
(WebKit::WebExtensionAction::clearCustomizations):
(WebKit::WebExtensionAction::setPopupPath):
(WebKit::WebExtensionAction::setLabel):
(WebKit::WebExtensionAction::setBadgeText):
(WebKit::WebExtensionAction::setEnabled):
* Source/WebKit/UIProcess/Extensions/Cocoa/WebExtensionContextCocoa.mm:
(WebKit::WebExtensionContext::postAsyncNotification):
(WebKit::WebExtensionContext::grantPermissionMatchPatterns):
(WebKit::WebExtensionContext::denyPermissionMatchPatterns):
(WebKit::WebExtensionContext::removeGrantedPermissionMatchPatterns):
(WebKit::WebExtensionContext::removeDeniedPermissionMatchPatterns):
(WebKit::WebExtensionContext::permissionState):
(WebKit::WebExtensionContext::setPermissionState):
(WebKit::WebExtensionContext::didChangeTabProperties):
(WebKit::WebExtensionContext::didStartProvisionalLoadForFrame):
(WebKit::WebExtensionContext::didCommitLoadForFrame):
(WebKit::WebExtensionContext::didFinishLoadForFrame):
(WebKit::WebExtensionContext::didFailLoadForFrame):
(WebKit::WebExtensionContext::performAction):
(WebKit::WebExtensionContext::performCommand):
(WebKit::WebExtensionContext::userGesturePerformed):
(WebKit::WebExtensionContext::hasActiveUserGesture const):
(WebKit::WebExtensionContext::clearUserGesture):
(WebKit::WebExtensionContext::loadBackgroundPageListenersFromStorage):
(WebKit::WebExtensionContext::saveBackgroundPageListenersToStorage):
(WebKit::WebExtensionContext::removeInjectedContent):
* Source/WebKit/UIProcess/Extensions/Cocoa/WebExtensionControllerCocoa.mm:
(WebKit::WebExtensionController::didStartProvisionalLoadForFrame):
(WebKit::WebExtensionController::didCommitLoadForFrame):
(WebKit::WebExtensionController::didFinishLoadForFrame):
(WebKit::WebExtensionController::didFailLoadForFrame):
* Source/WebKit/UIProcess/Extensions/Cocoa/WebExtensionTabCocoa.mm:
(WebKit::WebExtensionTab::changedParameters const):
* Source/WebKit/UIProcess/Extensions/WebExtensionAction.h:
* Source/WebKit/UIProcess/Extensions/WebExtensionContext.h:
(WebKit::WebExtensionContext::didChangeTabProperties):
* Source/WebKit/UIProcess/Extensions/WebExtensionController.h:
* Source/WebKit/UIProcess/Extensions/WebExtensionTab.h:
(WebKit::WebExtensionTab::changedParameters):
(WebKit::WebExtensionTab::hasActiveUserGesture):
(WebKit::WebExtensionTab::setActiveUserGesture):
(WebKit::WebExtensionTab::temporaryPermissionMatchPattern):
(WebKit::WebExtensionTab::setTemporaryPermissionMatchPattern):
(WebKit::WebExtensionTab::changedProperties const):
(WebKit::WebExtensionTab::addChangedProperties):
(WebKit::WebExtensionTab::clearChangedProperties):
* Tools/TestWebKitAPI/Tests/WebKitCocoa/WKWebExtensionAPIAction.mm:
(TestWebKitAPI::TEST):
* Tools/TestWebKitAPI/Tests/WebKitCocoa/WKWebExtensionAPITabs.mm:
(TestWebKitAPI::TEST):
* Tools/TestWebKitAPI/cocoa/WebExtensionUtilities.mm:
(-[TestWebExtensionTab webView:didCommitNavigation:]):

Canonical link: <a href="https://commits.webkit.org/270722@main">https://commits.webkit.org/270722@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/b75c18531727ac89fc283a6eb1b6ee89f336a595

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/26079 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/4689 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/27353 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/28175 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/23880 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/26403 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/6454 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/51/builds/2107 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/23930 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/26334 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/3546 "Passed tests") | [⏳ 🧪 api-mac](https://ews-build.webkit.org/#/builders/API-Tests-macOS-EWS "Waiting to run tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/28751 "Built successfully") | 
| | [⏳ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/iOS-17-Simulator-WPT-WK2-Tests-EWS "Waiting to run tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/23406 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/29459 "Passed tests") | 
| | [⏳ 🧪 api-ios](https://ews-build.webkit.org/#/builders/API-Tests-iOS-Simulator-EWS "Waiting to run tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/23787 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/27342 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/3205 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/50/builds/1405 "Passed tests") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/4603 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/3667 "Built successfully") | | | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/3375 "Built successfully and passed tests") | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/3526 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

b-weinstein commented on Source/WebKit/UIProcess/Extensions/Cocoa/WebExtensionContextCocoa.mm:
```diff
@@ -609,11 +609,21 @@ static _WKWebExtensionContextError toAPI(WebExtensionContext::Error error)
         firePermissionsEventListenerIfNecessary(WebExtensionEventListenerType::PermissionsOnRemoved, { }, matchPatterns);
 
+    // Fire the tab updated event for any tabs that match the changed patterns,
+    // now that the extension has / does not have permission to see the URL and title.
```
Super-nit: This could probably be on one line.

b-weinstein commented on Source/WebKit/UIProcess/Extensions/Cocoa/WebExtensionContextCocoa.mm:
```diff
@@ -1625,16 +1653,112 @@ static _WKWebExtensionContextError toAPI(WebExtensionContext::Error error)
+    auto tab = getTab(pageID);
+
+    // Dispatch webNavigation events.
+    if (tab && hasPermission(_WKWebExtensionPermissionWebNavigation) && hasPermission(targetURL)) {
```
Do we want to pass tab to `hasPermission` so activeTab checks will happen? Or does that happen even without `tab` being passed?

xeenon replied:
Passing `tab` for the feature `hasPermission` is only needed for `tabs` permission, but I might as well do it here just for coverage in case that ever changes!

xeenon replied:
Oh, but you mean `hasPermission` for the URL! Yes, that is needed to work with `activeTab`.

b-weinstein commented on Source/WebKit/UIProcess/Extensions/Cocoa/WebExtensionContextCocoa.mm:
```diff
@@ -1625,16 +1653,112 @@ static _WKWebExtensionContextError toAPI(WebExtensionContext::Error error)
+    }
+
+    // Dispatch webNavigation events.
+    if (tab && hasPermission(_WKWebExtensionPermissionWebNavigation) && hasPermission(frameURL)) {
```
Ditto passing `tab` to `hasPermission`.

b-weinstein commented on Source/WebKit/UIProcess/Extensions/Cocoa/WebExtensionContextCocoa.mm:
```diff
@@ -1625,16 +1653,112 @@ static _WKWebExtensionContextError toAPI(WebExtensionContext::Error error)
+    auto tab = getTab(pageID);
+
+    // Dispatch webNavigation events.
+    if (tab && hasPermission(_WKWebExtensionPermissionWebNavigation) && hasPermission(frameURL)) {
```
Ditto.

b-weinstein commented on Source/WebKit/UIProcess/Extensions/Cocoa/WebExtensionContextCocoa.mm:
```diff
@@ -1625,16 +1653,112 @@ static _WKWebExtensionContextError toAPI(WebExtensionContext::Error error)
+    auto tab = getTab(pageID);
+
+    // Dispatch webNavigation events.
+    if (tab && hasPermission(_WKWebExtensionPermissionWebNavigation) && hasPermission(frameURL)) {
```
Ditto.

xeenon committed:
Clear activeTab and action tab customization when tab navigates.
https://webkit.org/b/263918
rdar://problem/117704349

Reviewed by Brian Weinstein.

Moves the webNavigation load events to WebExtensionContext, since more things will
need to tie into these hooks.

Fixes injected style removal when webPageID is used, so it only removes the styles
when the page they apply to is navigated. Also this was incorrectly behind the
webNavigation permission checks before, when it needs to always happen.

Properly clears the activeTb permission on tab navigation, and fires the onUpdated
events for URL and Title when activeTab comes and goes, as well as any pattern grant.
This matches what Safari currently does when permissions are granted.

Clears the action customizations if the tab navigates. This matches the other browsers
and what Safari currently does.

Fixes the match pattern notifications to use the right constant.

* Source/WebKit/UIProcess/Extensions/Cocoa/WebExtensionActionCocoa.mm:
(WebKit::WebExtensionAction::clearCustomizations):
(WebKit::WebExtensionAction::setPopupPath):
(WebKit::WebExtensionAction::setLabel):
(WebKit::WebExtensionAction::setBadgeText):
(WebKit::WebExtensionAction::setEnabled):
* Source/WebKit/UIProcess/Extensions/Cocoa/WebExtensionContextCocoa.mm:
(WebKit::WebExtensionContext::postAsyncNotification):
(WebKit::WebExtensionContext::grantPermissionMatchPatterns):
(WebKit::WebExtensionContext::denyPermissionMatchPatterns):
(WebKit::WebExtensionContext::removeGrantedPermissionMatchPatterns):
(WebKit::WebExtensionContext::removeDeniedPermissionMatchPatterns):
(WebKit::WebExtensionContext::permissionState):
(WebKit::WebExtensionContext::setPermissionState):
(WebKit::WebExtensionContext::didChangeTabProperties):
(WebKit::WebExtensionContext::didStartProvisionalLoadForFrame):
(WebKit::WebExtensionContext::didCommitLoadForFrame):
(WebKit::WebExtensionContext::didFinishLoadForFrame):
(WebKit::WebExtensionContext::didFailLoadForFrame):
(WebKit::WebExtensionContext::performAction):
(WebKit::WebExtensionContext::performCommand):
(WebKit::WebExtensionContext::userGesturePerformed):
(WebKit::WebExtensionContext::hasActiveUserGesture const):
(WebKit::WebExtensionContext::clearUserGesture):
(WebKit::WebExtensionContext::loadBackgroundPageListenersFromStorage):
(WebKit::WebExtensionContext::saveBackgroundPageListenersToStorage):
(WebKit::WebExtensionContext::removeInjectedContent):
* Source/WebKit/UIProcess/Extensions/Cocoa/WebExtensionControllerCocoa.mm:
(WebKit::WebExtensionController::didStartProvisionalLoadForFrame):
(WebKit::WebExtensionController::didCommitLoadForFrame):
(WebKit::WebExtensionController::didFinishLoadForFrame):
(WebKit::WebExtensionController::didFailLoadForFrame):
* Source/WebKit/UIProcess/Extensions/Cocoa/WebExtensionTabCocoa.mm:
(WebKit::WebExtensionTab::changedParameters const):
* Source/WebKit/UIProcess/Extensions/WebExtensionAction.h:
* Source/WebKit/UIProcess/Extensions/WebExtensionContext.h:
(WebKit::WebExtensionContext::didChangeTabProperties):
* Source/WebKit/UIProcess/Extensions/WebExtensionController.h:
* Source/WebKit/UIProcess/Extensions/WebExtensionTab.h:
(WebKit::WebExtensionTab::changedParameters):
(WebKit::WebExtensionTab::hasActiveUserGesture):
(WebKit::WebExtensionTab::setActiveUserGesture):
(WebKit::WebExtensionTab::temporaryPermissionMatchPattern):
(WebKit::WebExtensionTab::setTemporaryPermissionMatchPattern):
(WebKit::WebExtensionTab::changedProperties const):
(WebKit::WebExtensionTab::addChangedProperties):
(WebKit::WebExtensionTab::clearChangedProperties):
* Tools/TestWebKitAPI/Tests/WebKitCocoa/WKWebExtensionAPIAction.mm:
(TestWebKitAPI::TEST):
* Tools/TestWebKitAPI/Tests/WebKitCocoa/WKWebExtensionAPITabs.mm:
(TestWebKitAPI::TEST):
* Tools/TestWebKitAPI/cocoa/WebExtensionUtilities.mm:
(-[TestWebExtensionTab webView:didCommitNavigation:]):

Canonical link: https://commits.webkit.org/270722@main
